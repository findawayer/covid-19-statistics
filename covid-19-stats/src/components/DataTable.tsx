import type { FunctionComponent, MouseEvent, TouchEvent } from 'react';
import React from 'react';

import { formatValue } from '../utils';
import {
  StyledBody,
  StyledCell,
  StyledHead,
  StyledHeadCell,
  StyledTable
} from './styles/DataTable';

type ColumnDefinition = {
  key: string;
  alias: string;
};

// Row --------------------

interface DataTableRowProps {
  item: TODO;
  columns: readonly ColumnDefinition[];
}

const DataTableRow: FunctionComponent<DataTableRowProps> = ({
  item,
  columns
}) => {
  return (
    <>
      {columns.map(({ key }) => (
        <StyledCell key={key}>{formatValue(item[key])}</StyledCell>
      ))}
    </>
  );
};

// Table --------------------

interface DataTableProps {
  data: TODO[] | null;
  columns: readonly ColumnDefinition[];
  handleColumnClick?: (
    sortKey: string
  ) => (event: MouseEvent | TouchEvent) => void;
}

const DataTable: FunctionComponent<DataTableProps> = ({
  data,
  columns,
  handleColumnClick
}) => {
  return (
    <StyledTable>
      <StyledHead>
        {columns.map(({ key, alias }) => (
          <StyledHeadCell
            key={key}
            onClick={handleColumnClick ? handleColumnClick(key) : undefined}
          >
            {alias ?? key}
          </StyledHeadCell>
        ))}
      </StyledHead>
      <StyledBody>
        {data &&
          data.map(item => (
            <DataTableRow key={item.country} item={item} columns={columns} />
          ))}
      </StyledBody>
    </StyledTable>
  );
};

export default DataTable;
