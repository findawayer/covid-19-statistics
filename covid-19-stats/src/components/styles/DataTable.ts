import styled from '@emotion/styled';

import ChevronDown from '../svg/chevron-down.svg';

export const StyledTable = styled.div`
  position: relative;
  min-width: 40em;
`;

export const StyledCell = styled.div`
  padding: 0.75em 1em;
  border: solid #ccc;
  border-width: 0 0 1px 1px;
`;

export const StyledHeadCell = styled(StyledCell)`
  position: relative;
  padding-top: 1em;
  padding-bottom: 1em;
  background-color: #eee;
  font-weight: 700;
  cursor: pointer;

  &.is-sort-key::after {
    content: '';
    position: absolute;
    top: 0;
    right: 1em;
    bottom: 0;
    display: block;
    width: 1rem;
    height: 1rem;
    margin: auto;
    background: url(${ChevronDown}) no-repeat center / 100% auto;
  }

  &.is-ascending::after {
    transform: rotate(-0.5turn);
  }
`;

export const StyledRowGroup = styled.div`
  display: grid;
  grid-template-columns: repeat(5, 1fr);
  box-sizing: border-box;
  background: #fff;
  border: solid #ccc;
  border-width: 1px 1px 0 0;
`;

export const StyledHead = styled(StyledRowGroup)`
  position: sticky;
  top: 0;
`;

export const StyledBody = styled(StyledRowGroup)`
  margin-top: -1px;
`;
