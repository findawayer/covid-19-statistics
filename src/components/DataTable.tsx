import type { FunctionComponent, MouseEvent, TouchEvent } from 'react';
import React from 'react';

import { formatValue } from '../utils';
import {
  StyledCell,
  StyledHeadCell,
  StyledSortIcon,
  StyledTable,
} from './styles/DataTable';

type ColumnDefinition = {
  key: string;
  alias: string;
};

// Row --------------------

interface DataTableRowProps {
  item: TODO;
  columns: readonly ColumnDefinition[];
}

const DataTableRow: FunctionComponent<DataTableRowProps> = ({
  item,
  columns,
}) => {
  return (
    <tr>
      {columns.map(({ key }) => (
        <StyledCell key={key}>{formatValue(item[key])}</StyledCell>
      ))}
    </tr>
  );
};

// Table --------------------

interface DataTableProps {
  data: TODO[] | null;
  columns: readonly ColumnDefinition[];
  idKey: string;
  sortKey?: string | null;
  sortDirection?: 'ascending' | 'descending';
  handleColumnClick?: (
    sortKey: string,
  ) => (event: MouseEvent | TouchEvent) => void;
}

const DataTable: FunctionComponent<DataTableProps> = ({
  data,
  columns,
  idKey,
  sortKey,
  sortDirection,
  handleColumnClick,
}) => {
  return (
    <StyledTable>
      <thead data-testid="thead">
        <tr>
          {columns.map(({ key, alias }) => (
            <StyledHeadCell
              key={key}
              onClick={handleColumnClick ? handleColumnClick(key) : undefined}
              data-testid={key}
            >
              {alias ?? key}
              <StyledSortIcon
                isActive={key === sortKey}
                isAscending={sortDirection === 'ascending'}
              />
            </StyledHeadCell>
          ))}
        </tr>
      </thead>
      <tbody data-testid="tbody">
        {data &&
          data.map(item => (
            <DataTableRow key={item[idKey]} item={item} columns={columns} />
          ))}
      </tbody>
    </StyledTable>
  );
};

export default DataTable;
