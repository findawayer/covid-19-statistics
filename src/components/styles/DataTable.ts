import styled from '@emotion/styled';

import ChevronDown from '../svg/chevron-down.svg';

export const StyledTable = styled.table`
  position: relative;
  width: 100%;
  border-collapse: separate;
  border-spacing: 0;
`;

export const StyledCell = styled.td`
  padding: 0.75em 1em;
  border: double #ccc;
  border-width: 0 1px 1px 0;
  text-align: left;

  &:last-child {
    border-right-width: 0;
  }
`;

// Some browsers only support sticky on `th`, not on `thead`.
export const StyledHeadCell = styled(StyledCell)`
  position: sticky;
  top: 0;
  padding-top: 1em;
  padding-bottom: 1em;
  border-top-width: 1px;
  background-color: #eee;
  font-weight: 700;
  cursor: pointer;
`.withComponent('th');

type StyledSortIconProps = {
  isActive: boolean;
  isAscending: boolean;
};

export const StyledSortIcon = styled.i<StyledSortIconProps>`
  width: 1rem;
  height: 1rem;
  float: right;
  visibility: ${({ isActive }) => (isActive ? 'visible' : 'hidden')};
  background: url(${ChevronDown}) no-repeat center / 100% auto;
  ${({ isAscending }) => isAscending && `transform: rotate(-0.5turn);`}
`;
