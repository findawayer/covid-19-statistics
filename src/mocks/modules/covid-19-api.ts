import type { CovidCountryStat } from '../../models/covid-19-api';

export const covidCountryStatMock: CovidCountryStat[] = [
  {
    code: 'KR',
    confirmed: 68664,
    country: 'South Korea',
    critical: 401,
    deaths: 1125,
    lastChange: '2021-01-10T06:16:39+01:00',
    lastUpdate: '2021-01-10T09:30:04+01:00',
    latitude: 35.907757,
    longitude: 127.766922,
    recovered: 50409,
  },
  {
    code: 'MN',
    confirmed: 1429,
    country: 'Mongolia',
    critical: 14,
    deaths: 2,
    lastChange: '2021-01-10T06:24:51+01:00',
    lastUpdate: '2021-01-10T09:30:04+01:00',
    latitude: 46.862496,
    longitude: 103.846656,
    recovered: 896,
  },
  {
    code: 'TW',
    confirmed: 828,
    country: 'Taiwan',
    critical: 0,
    deaths: 7,
    lastChange: '2021-01-09T07:15:52+01:00',
    lastUpdate: '2021-01-10T09:30:04+01:00',
    latitude: 23.69781,
    longitude: 120.960515,
    recovered: 720,
  },
];
