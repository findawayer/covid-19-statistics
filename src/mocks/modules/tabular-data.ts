export const tabularData = [
  { name: 'foo', size: 10 },
  { name: 'bar', size: 500 },
  { name: 'baz', size: 1000 },
  { name: 'qux', size: 1000000 },
];

export const tableColumns = [
  { key: 'name', alias: '이름' },
  { key: 'size', alias: '크기' },
];
