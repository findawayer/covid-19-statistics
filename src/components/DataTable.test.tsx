import { render } from '@testing-library/react';

import DataTable from './DataTable';
import { tableColumns, tabularData } from '../mocks/modules/tabular-data';

describe('<DataTable />', () => {
  it('should render data as table', () => {
    const handleColumnClick = jest.fn();
    const { container } = render(
      <DataTable
        data={tabularData}
        columns={tableColumns}
        idKey="name"
        sortKey="name"
        sortDirection="descending"
        handleColumnClick={handleColumnClick}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
