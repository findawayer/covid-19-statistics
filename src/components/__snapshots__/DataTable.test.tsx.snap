// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataTable /> should render data as table 1`] = `
<div>
  <table
    class="css-1ivotix"
  >
    <thead
      data-testid="thead"
    >
      <tr>
        <th
          class="css-rc0p4j"
          data-testid="name"
        >
          이름
          <i
            class="css-trwj99"
          />
        </th>
        <th
          class="css-rc0p4j"
          data-testid="size"
        >
          크기
          <i
            class="css-fmrrrd"
          />
        </th>
      </tr>
    </thead>
    <tbody
      data-testid="tbody"
    >
      <tr>
        <td
          class="css-1g8ku9a"
        >
          foo
        </td>
        <td
          class="css-1g8ku9a"
        >
          10
        </td>
      </tr>
      <tr>
        <td
          class="css-1g8ku9a"
        >
          bar
        </td>
        <td
          class="css-1g8ku9a"
        >
          500
        </td>
      </tr>
      <tr>
        <td
          class="css-1g8ku9a"
        >
          baz
        </td>
        <td
          class="css-1g8ku9a"
        >
          1,000
        </td>
      </tr>
      <tr>
        <td
          class="css-1g8ku9a"
        >
          qux
        </td>
        <td
          class="css-1g8ku9a"
        >
          1,000,000
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
